### PR TITLE
Proofs for Arithmetic

### PR DESCRIPTION
**Pull request by alex-ozdemir**
Includes a few infrastructural changes too:
* allow assumption/scope shadowing in `ProofNode`s
* allow assumptions like `(= false (= a b))` when closing `(not (= b a))` in `mkProof`.

We do not provide proofs for all of arithmetic. Notably we omit
* black-box conflicts
* a number of arithmetic lemmas.

Handling these is future work.

### CODE REVIEW

File: src/theory/arith/congruence_manager.h [hunk 1]
```diff
@@ -101,6 +102,16 @@ class ArithCongruenceManager {
 
   eq::EqualityEngine d_ee;
 
+  /** proof manager */
+  ProofNodeManager* d_pnm;
+  // A proof generator for storing proofs of facts that are asserted to the EQ
```
Comment (ajreynol) [inline src/theory/arith/congruence_manager.h:107]: I believe we always use comments enclosed by `/** */` for documentation in h files.

File: src/theory/arith/congruence_manager.h [hunk 2]
```diff
@@ -129,8 +140,26 @@ class ArithCongruenceManager {
   bool propagate(TNode x);
   void explain(TNode literal, std::vector<TNode>& assumptions);
 
+  // Assert this literal to the eq engine. Common functionality for
```
Comment (ajreynol) [inline src/theory/arith/congruence_manager.h:143]: Ditto.

File: src/theory/arith/constraint.h [hunk 3]
```diff
@@ -473,6 +476,10 @@ class Constraint {
     return d_literal;
   }
 
+  // Gets a literal in the normal form suitable for proofs.
+  // That is, (sum of non-const monomials) >< const.
+  Node getProofLiteral() const;
```
Comment (ajreynol) [inline src/theory/arith/constraint.h:481]: This is basically converting the constraint to a Node?  Would a better name be `toProofLiteral()`?

File: src/theory/arith/theory_arith_private.cpp [hunk 4]
```diff
@@ -4684,55 +4745,116 @@ bool TheoryArithPrivate::rowImplicationCanBeApplied(RowIndex ridx, bool rowUp, C
...
+        farkasCoefficients.reserve(coeffs->size());
+        auto nm = NodeManager::currentNM();
+        std::transform(
+            coeffs->begin(),
+            coeffs->end(),
+            std::back_inserter(farkasCoefficients),
+            [nm](const Rational& r) { return nm->mkConst<Rational>(r); });
+
+        // Prove bottom.
+        auto sumPf = d_pnm->mkNode(
```
Comment (ajreynol) [inline src/theory/arith/theory_arith_private.cpp:4831]: nit: Generally I prefer not using auto (IMHO I find types make things easier to read).

File: src/theory/arith/congruence_manager.h [hunk 5]
```diff
@@ -101,6 +102,16 @@ class ArithCongruenceManager {
...
   eq::EqualityEngine d_ee;
 
+  /** proof manager */
+  ProofNodeManager* d_pnm;
+  // A proof generator for storing proofs of facts that are asserted to the EQ
+  // engine. Note that these proofs **are not closed**, and assume the
+  // explanation of these facts. This is why this generator is separate from the
+  // TheoryArithPrivate generator, which stores closed proofs.
+  std::unique_ptr<EagerProofGenerator> d_pfGenEe;
+  // A proof generator for TrustNodes sent to the theory.
```
Comment (ajreynol) [inline src/theory/arith/congruence_manager.h:112]: I'd be curious to see more documentation on this.  Do you mean TrustNodes sent to this class? (the main Theory class doesn't input TrustNode).

File: src/theory/arith/constraint.h (hunk 3 above)
Comment (alex-ozdemir) [reply src/theory/arith/constraint.h:481]: The name mirrors `getLiteral`, a method which returns the externally-valid literal associated with this constraint.

I like the connection between the names, but I'm open to changing it.

Comment (ajreynol) [reply src/theory/arith/constraint.h:481]: Ah I see, that's fine then.

File: src/expr/proof_node.cpp [hunk 6]
```diff
@@ -54,8 +54,10 @@ void ProofNode::getFreeAssumptionsMap(
   std::unordered_map<ProofNode*, bool> visited;
   std::unordered_map<ProofNode*, bool>::iterator it;
   std::vector<ProofNode*> visit;
-  // the current set of formulas bound by SCOPE
-  std::unordered_set<Node, NodeHashFunction> currentScope;
+  // Maps an bound assumption to the number of bindings it is under
```
Comment (HanielB) [inline src/expr/proof_node.cpp:57]: a bound

File: src/expr/proof_node.cpp [hunk 7]
```diff
@@ -54,8 +54,10 @@ void ProofNode::getFreeAssumptionsMap(
...
+  // e.g. in (SCOPE (SCOPE (ASSUME x) (x y)) (y)), y would be mapped to 2 at
+  // (ASSUME x), and x would be mapped to 1.
+  std::unordered_map<Node, uint32_t, NodeHashFunction> scopeDepth;
```
Comment (HanielB) [inline src/expr/proof_node.cpp:60]: Could you say how the scopeDepth is used in the code below, hopefully expanding the example above?

File: src/expr/proof_node.cpp [hunk 8]
```diff
@@ -104,7 +104,16 @@ void ProofNode::getFreeAssumptionsMap(
...
+        auto scopeCt = scopeDepth.find(a);
+        Assert(scopeCt != scopeDepth.end());
+        if (scopeCt->second == 1)
+        {
+          scopeDepth.erase(scopeCt);
+        }
+        else
+        {
+          scopeCt->second -= 1;
+        }
```
Comment (HanielB) [inline src/expr/proof_node.cpp:116]: You could refactor this to always decrement and check if 0, in which case you erase.

File: src/expr/proof_node_manager.cpp [hunk 9]
```diff
@@ -134,34 +135,60 @@ std::shared_ptr<ProofNode> ProofNodeManager::mkScope(
...
+      {
+        Trace("pnm-scope") << "- introduce explicit predicate " << aPred
+                           << " via " << (symm ? "[reoriented] " : "") << a
+                           << " for " << fa.second.size() << " proof nodes"
+                           << std::endl;
+        std::shared_ptr<ProofNode> pfaa = mkAssume(flipPred ? symPred : aPred);
+        // Potentially apply symmetry to the real assumption
+        if (flipPred) {
+          pfaa = mkNode(PfRule::SYMM, pfaa, {}, aPred);
+        }
```
Comment (HanielB) [inline src/expr/proof_node_manager.cpp:170]: clang-format

File: src/expr/proof_node_manager.cpp [hunk 10]
```diff
@@ -134,34 +135,60 @@ std::shared_ptr<ProofNode> ProofNodeManager::mkScope(
...
+        // justification of the assumption
+        if (symm)
+        {
+          pfaa = mkNode(pol ? PfRule::TRUE_INTRO : PfRule::FALSE_INTRO,
+                        pfaa,
+                        {},
+                        a[1].eqNode(a[0]));
+        }
+        PfRule updateRule =
+            symm ? PfRule::SYMM
```
Comment (HanielB) [inline src/expr/proof_node_manager.cpp:181]: Now that we have two kind of symmetries for this case it'd be good to rename this variable. Following the new one you created we could call this `flipEqualityPredIntro`.

File: src/theory/arith/congruence_manager.h (hunk 1 above)
Comment (alex-ozdemir) [reply src/theory/arith/congruence_manager.h:107]: Done in 8e836b883.

File: src/theory/arith/congruence_manager.h (hunk 2 above)
Comment (alex-ozdemir) [reply src/theory/arith/congruence_manager.h:143]: Done in 8e836b883.

File: src/theory/arith/congruence_manager.h (hunk 5 above)
Comment (alex-ozdemir) [reply src/theory/arith/congruence_manager.h:112]: Done in ef7ba20c3.

File: src/expr/proof_node.cpp (hunk 6 above)
Comment (alex-ozdemir) [reply src/expr/proof_node.cpp:57]: Done in 8ee9a9a13.

File: src/expr/proof_node.cpp (hunk 7 above)
Comment (alex-ozdemir) [reply src/expr/proof_node.cpp:60]: Done in 8ee9a9a13.

File: src/expr/proof_node.cpp (hunk 8 above)
Comment (alex-ozdemir) [reply src/expr/proof_node.cpp:116]: Done in 26dbf02cd.

File: src/expr/proof_node_manager.cpp (hunk 10 above)
Comment (alex-ozdemir) [reply src/expr/proof_node_manager.cpp:181]: Done in 4fe9d7bee

File: src/expr/proof_node_manager.cpp (hunk 9 above)
Comment (alex-ozdemir) [reply src/expr/proof_node_manager.cpp:170]: Done in 407177abe and 22a21abfd.

**Pull request merged**